src/consensus.rs

```diff
@@ -93,37 +94,50 @@ impl<T: Proposition> Consensus<T> {
     // membership: gen = pending_gen
     /// Handles a signed vote
     /// Returns the vote we cast and the reached consensus vote in case consensus was reached
-    pub fn handle_signed_vote(
-        &mut self,
-        signed_vote: SignedVote<T>,
-        gen: Generation,
-    ) -> Result<VoteResponse<T>> {
+    pub fn handle_signed_vote(&mut self, signed_vote: SignedVote<T>) -> Result<VoteResponse<T>> {
+        info!("[MBR-{}] handling vote {:?}", self.id(), signed_vote);
         if let Err(faults) = self.detect_byzantine_voters(&signed_vote) {
-            println!("[MBR-{}] Found faults {:?}", self.id(), faults);
+            info!("[MBR-{}] Found faults {:?}", self.id(), faults);
             self.faults.extend(faults);
         }
+        if self.faults.contains_key(&signed_vote.voter) {
+            info!("[MBR-{}] dropping vote from faulty voter", self.id());
+            return Ok(VoteResponse::WaitingForMoreVotes);
+        }
+        if !self.is_new_vote_from_voter(&signed_vote) {
+            info!("[MBR-{}] dropping old vote {:?}", self.id(), signed_vote);
+            return Ok(VoteResponse::WaitingForMoreVotes);
+        }
 
         let their_decision = self.get_super_majority_over_super_majorities(
             &signed_vote.unpack_votes().into_iter().cloned().collect(),
         )?;
 
         if let Some(decision) = self.decision.clone() {
-            if their_decision.is_none() && self.is_new_vote_from_voter(&signed_vote) {
-                println!(
+            let resp = if their_decision.is_none() {
+                info!(
                     "[MBR-{}] We've already terminated, responding with decision",
                     self.id()
                 );
-                return Ok(VoteResponse::Broadcast(
-                    self.build_super_majority_vote(decision.votes, gen)?,
-                ));
-            }
+                let vote = self.build_super_majority_vote(
+                    decision.votes,
+                    signed_vote.vote.gen,
+                    &BTreeSet::from_iter(decision.faults.iter().map(Fault::voter_at_fault)),
+                )?;
```

Comment:
This was a bug: Previously, `build_super_majority` would use `self.faults()` to populate the `SuperMajority` vote, but this is a problem if we detect a new fault after we've decided.

If a new fault is added to the super-majority vote, then the votes from that faulty node will be filtered out, which may reduce us to below super-majority and change the end decision. 

So the fix is to use the faults at the time of the decision.